Complete commit of complex numbers

gordonguthrie opened the pull request:
Complex numbers are now in the lexer (with a couple of additional hand-written pre-parsing steps

Need to be added to the actual parser.

A full set of tests and edge cases and technical documentation is prepared waiting on getting the full tests working.

Commit:
minor tidy up

hypernumbers committed:
Added complex numbers to the lexer/parser

These changes have been implemented as a pair of hand-descending lookahead transforms to pick up fragments of possible variable names that match patterns of complex numbers and restructure the token stream.

In theory this could all be done in the parser - but at some cost:

* bloat and confusion in the parser
* intertwingling of test issue between handling complex numbers and everthing else

It kinda made sense from a developmental perspective to drive it through this way with its own lexer test suite and make sure I understood and caught all the edge cases without eleventy six million existing tests going beskerkers

cannadayr commented:
One quick note: In APL, dyadic `|` is the modulo operator. I believe in erlang it is `rem`.

In APL (tested the following on GNU APL and Dyalog), it does operate on imaginary numbers:
```
      1J1|1J2
¯1
      1J1|1J3
0
      1J2|1J3
0J1
```

Not immediately clear what equation it is performing to arrive at this.

cannadayr commented:
https://aplwiki.com/wiki/Residue#Caveats

hypernumbers committed:
test for failing and bug fixd

hypernumbers committed:
Big mix of changes.

At the core is a big implementation of complex numbers:

* the basis of scalar extensions
* how we can extend primitives to tupes, records and maps

As part of this there is a host of other minor but important work:

* rename of liffey record to ast as we are not using LFE anymore ;-)
* short form macros to make code more readable
* raw expressions are considered acceptable - they return themselves.

The main non-functional thing is work on making a better developer experience including some pages for developers and how to get started.

Part of this is more test shadowing and a description of the shadow structure that allows new developers to break loads of tests and be confident that they can fix me.

Structuring tests around shadowing lets the test suite tell you which broken tests to fix first when you get massive test failures and is critical to a happy developer experience.

Previously we had the core tests in `test/` which test the basics of the lexer/parser and then the bulk of the tests in 'test/generated_tests` from the documentation.

The generated tests are ***shadowed*** by the basic tests - if the basic tests fail then sure as hell the feature tests will fail.

We now add a new set of shadowing tests - not as primitive as those in `test/`. These basic tests are on the page `docs/implementation_reference/basics.md`. They check that the core runtime custom paths are correctly executed for each branch.

hypernumbers committed:
fix merge

hypernumbers committed:
fix merge

hypernumbers committed:
Merge branch 'master' into gordonguthrie_complex_numbers
